mirrord/intproxy/src/proxies/files.rs

```diff
@@ -0,0 +1,1345 @@
+        self.protocol_version.replace(version);
+    }
+
+    // #[tracing::instrument(level = Level::TRACE, skip(message_bus))]
+    async fn file_request(
+        &mut self,
+        request: FileRequest,
+        layer_id: LayerId,
+        message_id: MessageId,
+        message_bus: &mut MessageBus<Self>,
+    ) {
+        match request {
+            // Should trigger remote close only when the fd is closed in all layer instances.
+            FileRequest::Close(close) => {
+                if self.remote_files.remove(layer_id, close.fd) {
+                    self.files_data.remove(&close.fd);
+                    message_bus
+                        .send(ClientMessage::FileRequest(FileRequest::Close(close)))
+                        .await;
+                }
+            }
+
+            // Should trigger remote close only when the fd is closed in all layer instances.
+            FileRequest::CloseDir(close) => {
+                if self.remote_dirs.remove(layer_id, close.remote_fd) {
+                    self.dirs_data.remove(&close.remote_fd);
+                    message_bus
+                        .send(ClientMessage::FileRequest(FileRequest::CloseDir(close)))
+                        .await;
+                }
+            }
+
+            // May require storing additional data in the request queue.
+            FileRequest::Open(open) => {
+                let additional_data = (self.buffer_reads && open.open_options.is_read_only())
+                    .then_some(AdditionalRequestData::OpenBuffered)
+                    .unwrap_or_default();
+                self.request_queue
+                    .insert_with_data(message_id, layer_id, additional_data);
+                message_bus
+                    .send(ClientMessage::FileRequest(FileRequest::Open(open)))
+                    .await;
+            }
+
+            // May require storing additional data in the request queue.
+            FileRequest::OpenRelative(open) => {
+                let additional_data = (self.buffer_reads && open.open_options.is_read_only())
+                    .then_some(AdditionalRequestData::OpenBuffered)
+                    .unwrap_or_default();
+                self.request_queue
+                    .insert_with_data(message_id, layer_id, additional_data);
+                message_bus
+                    .send(ClientMessage::FileRequest(FileRequest::OpenRelative(open)))
+                    .await;
+            }
+
+            // Try to use local buffer if possible.
+            FileRequest::Read(read) => match self.files_data.get_mut(&read.remote_fd) {
+                // File is buffered.
+                Some(data) => {
+                    let from_buffer = data.read_from_buffer(read.buffer_size, data.fd_position);
+                    if let Some(from_buffer) = from_buffer {
+                        let bytes = from_buffer.to_vec();
+                        data.fd_position += read.buffer_size;
```

Comment:
`fd_position`? Do you mean that this is the cursor position in this buffer for the `fd` of the file we are reading?

Reply:
I mean position of the file descriptor in the file, simulated in here. Normally fd position of remote files is handled in the agent - the agent opens a raw OS file descriptor, position of which is managed by the kernel. Following this approach for buffered files was a bit problematic. So I moved managing fd position to the intproxy (see `File buffering` section of `FilesProxy` doc)